src/coprocessor/dag/rpn_expr/impl_arithmetic.rs

```diff
@@ -313,6 +314,93 @@ impl ArithmeticOp for DecimalMultiply {
     }
 }
 
+#[derive(Debug)]
+pub struct IntDivideInt;
+
+impl ArithmeticOp for IntDivideInt {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        if *rhs == 0 {
+            return Ok(None);
+        }
+        Ok(Some(div_i64(*lhs, *rhs)?))
+    }
+}
+
+#[derive(Debug)]
+pub struct IntDivideUint;
+
+impl ArithmeticOp for IntDivideUint {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        if *rhs == 0 {
+            return Ok(None);
+        }
+        Ok(Some(div_i64_with_u64(*lhs, *rhs as u64).map(|r| r as i64)?))
+    }
+}
+
+#[derive(Debug)]
+pub struct UintDivideUint;
+
+impl ArithmeticOp for UintDivideUint {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        if *rhs == 0 {
+            return Ok(None);
+        }
+        Ok(Some(((*lhs as u64) / (*rhs as u64)) as i64))
+    }
+}
+
+#[derive(Debug)]
+pub struct UintDivideInt;
+
+impl ArithmeticOp for UintDivideInt {
+    type T = Int;
+
+    fn calc(lhs: &Int, rhs: &Int) -> Result<Option<Int>> {
+        if *rhs == 0 {
+            return Ok(None);
+        }
+        Ok(Some(div_u64_with_i64(*lhs as u64, *rhs).map(|r| r as i64)?))
+    }
+}
+
+#[rpn_fn(capture = [ctx])]
+#[inline]
+fn int_divide_decimal(
+    ctx: &mut EvalContext,
+    lhs: &Option<Decimal>,
+    rhs: &Option<Decimal>,
+) -> Result<Option<Int>> {
+    use crate::coprocessor::codec::mysql::Res;
+
+    if lhs.is_none() || rhs.is_none() {
+        return Ok(None);
+    }
+    let lhs = lhs.as_ref().unwrap();
+    let rhs = rhs.as_ref().unwrap();
+
+    match lhs / rhs {
+        Some(v) => match v {
+            Res::Ok(v) => match v.as_i64() {
+                Res::Ok(v_i64) => Ok(Some(v_i64)),
+                Res::Truncated(v_i64) => Ok(Some(v_i64)),
+                Res::Overflow(_) => {
+                    Err(Error::overflow("BIGINT", &format!("({} / {})", lhs, rhs)))?
+                }
+            },
+            Res::Truncated(_) => Err(Error::truncated())?,
+            Res::Overflow(_) => Err(Error::overflow("DECIMAL", &format!("({} / {})", lhs, rhs)))?,
+        },
+        None => Ok(ctx.handle_division_by_zero().map(|()| None)?),
```

Comment:
Why not use `Decimal::is_zero` to check whether the `rhs` is zero?

Reply:
> Why not use `Decimal::is_zero` to check whether the `rhs` is zero?

I think Div<&'a Decimal> call `do_div_mod_impl`, which will handle `division_by_zero` for us. It's unnecessary to do any extra checking.